### DISCUSSION TITLE
Identify source of map data

### DISCUSSION THREAD

**Discussion by kerbylane** [dabreegster/abstreet#930] [Q&A]
I've just noticed a problem (and possibly another) with a lane at an intersection. I'm trying to the source of the data that leads to this. Is there a way to tell from the UI?

In Seattle at the intersection of Greenwood Ave N and N 105th St the westbound curb lane of 105th is listed as turn-only. In reality drivers can turn or go straight.

The other possible problems is that, on northbound Greenwood, 2 lanes are shown as being able to go straight or turn right. I don't *think* that the center lane can make a right turn there. If that were allowed, while the right-hand lane can go straight there's a high likelihood of a collision. On the other hand, there doesn't appear to be any signage indicating that the center lane must go straight or that the curb lane must turn. So this is confusing to me.

<img width="420" alt="Screen Shot 2022-05-29 at 2 12 06 PM" src="https://user-images.githubusercontent.com/4228228/170892741-2952dfca-7567-44aa-9b5e-43e143502233.png">
.

**Comment by dabreegster** [comment]
> I'm trying to the source of the data that leads to this. Is there a way to tell from the UI?

Yes! https://a-b-street.github.io/docs/tech/dev/index.html#development-tips describes some of this. You'd make sure dev mode is on (**ctrl+S** or `--dev`), click a lane, go to the debug tab. The OSM tags are listed there, and there's an "Open OSM way button" too:
![Screenshot from 2022-05-30 10-23-15](https://user-images.githubusercontent.com/1664407/170961232-dc837eaa-9be6-4399-8fec-41fde1eee256.png)
The tags shown in the UI might differ from real OSM slightly -- you'll notice things like `abst_sidewalks_inferred`.

> In Seattle at the intersection of Greenwood Ave N and N 105th St the westbound curb lane of 105th is listed as turn-only. In reality drivers can turn or go straight.

In the latest version (0.3.22), this looks different / correct:
![Screenshot from 2022-05-30 10-24-40](https://user-images.githubusercontent.com/1664407/170961708-585be0a5-ac5b-46a9-bdbe-7a4406333eb8.png)
You can click a lane, then press **z** to see the turns originating there. Not sure why your screenshot shows the lane there as super tiny; is that a manual edit?

> If that were allowed, while the right-hand lane can go straight there's a high likelihood of a collision.

You're correct. The OSM tagging here (https://www.openstreetmap.org/way/428080881) just says there's a leftmost turn lane, and doesn't say anything about the other 2. Because OSM rarely maps all of the "common sense" turn restrictions, A/B Street makes guesses. In this case, I think there are 2 right turns because N 105th has 2 lanes. But you're right that this then means the rightmost lane of Greenwood shouldn't be able to go straight!

https://github.com/a-b-street/abstreet/blob/master/map_model/src/make/turns.rs has the code for generating turns at one intersection. `make_vehicle_turns` first generates all possible cases, then there's filters like `remove_merging_turns` and stuff that checks turn restrictions from OSM. This code isn't great and the testing is minimal (https://github.com/a-b-street/abstreet/blob/d98f18bb186b33ff1c636412600b0c98450e1547/tests/src/main.rs#L48) right now. We've been carving out pieces from A/B Sreet like [osm2lanes](https://github.com/a-b-street/osm2lanes) and stuff to generate road/intersection geometry from OSM, and a (quite distant) next step there is to pull out this turn generation logic into a separate library, overhaul it, and write lots of tests. But if you were interested in diving into the current code and improving it, this bug would be a great place to start!
